_drafts/2022-01-25-quick-and-easy-custom-hooks.md

```diff
@@ -0,0 +1,219 @@
+}
+```
+
+### Calling the hook
+
+Now that we have extracted the hook, we can call it from our component. This should just be a matter of removing the `useState` and `useEffect` logic, pulling `userId` from our custom hook call instead.
+
+Again, here's the diff from the old component implementation to the new one using our new custom hook:
+
+```diff
+const myComponent = () => {
+-  const [userId, setUserId] = useState(null)
++  const userId = useQueryParam()
+ 
+-  useEffect(() => {
+-    const urlParams = new URLSearchParams(window.location.search)
+-    const nextUserId = urlParams.get('user_id') || null
+-    setUserId(nextUserId)
+-  }, [window.location])
+ 
+  return userId ? <>Your user ID is: {userId}</> : <>No user ID found</>
+}
+```
+
+### Refactoring our hook for multiple uses
+
+As it stands, we're still just pulling `user_id` from our query parameters. Instead, it would be nice if we could pass in the parameter we're looking for, so that different components can use the same hook.
+
+To do this, we first need to accept a `param` argument for our hook: 
+
+
+```diff
+- const useQueryParam = () => {
++ const useQueryParam = (param) => {
+-  const [userId, setUserId] = useState(null)
++  const [paramValue, setParamValue] = useState(null)
+
+  useEffect(() => {
+    const urlParams = new URLSearchParams(window.location.search)
+-    const nextUserId = urlParams.get('user_id') || null
++    const nextParamValue = urlParams.get(param) || null
+-    setUserId(nextUserId)
++    setParamValue(nextParamValue)
+  }, [window.location])
+ 
+-  return userId
++  return paramValue
+}
+```
+
+Next, we can pass in the parameter from our component:
+
+```diff
+const myComponent = () => {
+-  const userId = useQueryParam()
++  const userId = useQueryParam('user_id')
+  return userId ? <>Your user ID is: {userId}</> : <>No user ID found</>
+}
+```
+
+
+### Putting it all together
+
+Altogether, we now have a very simple component and a custom hook with logic that can be reused elsewhere:
+
+```javascript
+const useQueryParam = (param) => {
+  const [paramValue, setParamValue] = useState(null)
+
+  useEffect(() => {
+    const urlParams = new URLSearchParams(window.location.search)
+    const nextParamValue = urlParams.get(param) || null
+    setParamValue(nextParamValue)
+  }, [window.location])
+
+  return paramValue
+}
+
+const myComponent = () => {
+  const userId = useQueryParam('user_id')
+  return userId ? <>Your user ID is: {userId}</> : <>No user ID found</>
+}
+```
+
+Notice very little has changed. This works functionally the same as before, but now we can call `useQueryParam` from multiple different components without needing to repeat code.
+
+
+
+## tl;dr
+
+To extract a custom hook from a complex component:
+1. Take all the logic above your component's JSX, and move it to a custom hook. 
+2. Return any data needed in your component from your custom hook.
+3. Call the custom hook from your component, and pull out whatever data is needed.
```

Comment:
The hook vs component repetition here is making it hard to understand quickly. May want number 2 to just say 'Return any data needed to render the JSX from your custom hook` but then 3 would be repetitive. Anyway, just think there's some repetition clean-up we could do here for clarity.

Reply:
That's a good point. I don't think 3 is necessarily repetitive after rewording 2. Curious what you think of the new wording?

https://github.com/himaxwell/himaxwell.github.io/pull/19/commits/9e18f83255c69c1ea146dc7184f3ba8931672d0d